hpx/lcos/local/channel.hpp

```diff
@@ -267,174 +534,293 @@ namespace hpx { namespace lcos { namespace local
+        using base_type::end;
+        using base_type::range;
     };
 
-    ///////////////////////////////////////////////////////////////////////////
+    // channel with a one-element buffer
     template <typename T>
-    class receive_channel
+    class one_element_channel : protected detail::channel_base<T>
     {
+        typedef detail::channel_base<T> base_type;
+
+    private:
+        friend class channel_iterator<T>;
+        friend class receive_channel<T>;
+        friend class send_channel<T>;
+
     public:
-        receive_channel(channel<T> const& c)
-          : channel_(c.channel_)
-        {}
+        typedef T value_type;
 
-        ///////////////////////////////////////////////////////////////////////
-        hpx::future<T> get(launch::async_policy,
-            std::size_t generation = std::size_t(-1)) const
-        {
-            return channel_->get(generation);
-        }
-        hpx::future<T> get(std::size_t generation = std::size_t(-1)) const
-        {
-            return get(launch::async, generation);
-        }
-        T get(launch::sync_policy, std::size_t generation = std::size_t(-1),
-            error_code& ec = throws) const
-        {
-            return channel_->get(generation, true).get(ec);
-        }
-        T get(launch::sync_policy, error_code& ec,
-            std::size_t generation = std::size_t(-1)) const
-        {
-            return channel_->get(generation, true).get(ec);
-        }
+        one_element_channel()
+          : base_type(new detail::one_element_channel<T>())
+        {}
 
-        ///////////////////////////////////////////////////////////////////////
-        channel_iterator<T> begin() const
-        {
-            return channel_iterator<T>(this);
-        }
-        channel_iterator<T> end() const
-        {
-            return channel_iterator<T>();
-        }
+        using base_type::get;
+        using base_type::set;
+        using base_type::close;
+        using base_type::begin;
+        using base_type::end;
+        using base_type::range;
+    };
 
-        channel_iterator<T> rbegin() const
-        {
-            return channel_iterator<T>(this);
-        }
-        channel_iterator<T> rend() const
-        {
-            return channel_iterator<T>();
-        }
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename T>
+    class receive_channel : protected detail::channel_base<T>
+    {
+        typedef detail::channel_base<T> base_type;
 
     private:
         friend class channel_iterator<T>;
+        friend class receive_channel<T>;
+        friend class send_channel<T>;
 
-    private:
-        boost::intrusive_ptr<detail::channel_base<T> > channel_;
+    public:
+        receive_channel(channel<T> const& c)
+          : base_type(c.get_channel_impl())
+        {}
+        receive_channel(one_element_channel<T> const& c)
+          : base_type(c.get_channel_impl())
+        {}
```

Comment:
No for two reasons:

- in that case a `receive_channel` would become constructible from a `send_channel`, for instance
- all channel classes are non-publicly derived  (using `protected` derivation) from `channel_base`